src/map/itemdb.cpp

```diff
@@ -2274,194 +2272,396 @@ bool itemdb_isNoEquip(struct item_data *id, uint16 m) {
 			}
-			data->script = code;
-		}
-		count++;
-	}
-	fclose(fp);
 
-	ShowStatus("Done reading '" CL_WHITE "%u" CL_RESET "' entries in '" CL_WHITE "%s" CL_RESET "'.\n", count, path);
+			if (this->nodeExists(optionNode, "Param")) {
+				int16 value;
 
-	return true;
-}
+				if (!this->asInt16(optionNode, "Param", value))
+					return 0;
 
-/**
- * Clear Item Random Option Group from memory
- * @author [Cydh]
- **/
-static int itemdb_randomopt_group_free(DBKey key, DBData *data, va_list ap) {
-	struct s_random_opt_group *g = (struct s_random_opt_group *)db_data2ptr(data);
-	if (!g)
-		return 0;
-	if (g->entries)
-		aFree(g->entries);
-	g->entries = NULL;
-	aFree(g);
-	return 1;
-}
+				entry->param = static_cast<int8>(value);
+			} else {
+				if (!exists)
+					entry->param = 0;
+			}
 
-/**
- * Get Item Random Option Group from itemdb_randomopt_group MapDB
- * @param id Random Option Group
- * @return Random Option Group data or NULL if not found
- * @author [Cydh]
- **/
-struct s_random_opt_group *itemdb_randomopt_group_exists(int id) {
-	return (struct s_random_opt_group *)uidb_get(itemdb_randomopt_group, id);
-}
+			if (this->nodeExists(optionNode, "Chance")) {
+				uint16 chance;
 
-/**
- * Read Item Random Option Group from db file
- * @author [Cydh]
- **/
-static bool itemdb_read_randomopt_group(char* str[], int columns, int current) {
-	int64 id_tmp;
-	int id = 0;
-	int i;
-	unsigned short rate = (unsigned short)strtoul(str[1], NULL, 10);
-	struct s_random_opt_group *g = NULL;
+				if (!this->asUInt16Rate(optionNode, "Chance", chance))
+					return 0;
 
-	if (!script_get_constant(str[0], &id_tmp)) {
-		ShowError("itemdb_read_randomopt_group: Invalid ID for Random Option Group '%s'.\n", str[0]);
-		return false;
-	}
+				entry->chance = chance;
+			} else {
+				if (!exists)
+					entry->chance = 0;
+			}
 
-	id = static_cast<int>(id_tmp);
+			entries.push_back(entry);
+		}
 
-	if ((columns-2)%3 != 0) {
-		ShowError("itemdb_read_randomopt_group: Invalid column entries '%d'.\n", columns);
-		return false;
+		randopt->slot.insert({ slot - 1, entries });
 	}
 
-	if (!(g = (struct s_random_opt_group *)uidb_get(itemdb_randomopt_group, id))) {
-		CREATE(g, struct s_random_opt_group, 1);
-		g->id = id;
-		g->total = 0;
-		g->entries = NULL;
-		uidb_put(itemdb_randomopt_group, g->id, g);
+	if (this->nodeExists(node, "MaxRandom")) {
+		uint16 max;
+
+		if (!this->asUInt16(node, "MaxRandom", max))
+			return 0;
+
+		randopt->max_random = max;
+	} else {
+		if (!exists)
+			randopt->max_random = 0;
 	}
 
-	RECREATE(g->entries, struct s_random_opt_group_entry, g->total + rate);
+	if (this->nodeExists(node, "Random")) {
+		if (randopt->slot.size() >= MAX_ITEM_RDM_OPT) {
+			this->invalidWarning(node["Random"], "No Random options can be added because there are no Random Option slots available.\n", MAX_ITEM_RDM_OPT);
+			return 0;
+		}
+
+		for (const YAML::Node &randomNode : node["Random"]) {
+			std::shared_ptr<s_random_opt_group_entry> entry;
```

Comment:
If I am not mistaken this code and the code above do the same thing and should be put into a sub function maybe?